Spatial_Engine/Query/IsFullyOnGrid.cs

```diff
@@ -0,0 +1,132 @@
+        [Input("element0D", "The IElement0D that will be checked for proximity to the grid.")]
+        [Input("grid", "The Grid to use for evaulation.")]
+        [Input("tolerance", "The maximum distance allowed from the Grid for this method to return true.", typeof(Length))]
+        [Output("isFullyOnGrid", "A boolean which is true if the geometrical representation of the IElement0D is fully within a set tolerance from the grid line.")]
+        public static bool IsFullyOnGrid(this IElement0D element0D, Grid grid, double tolerance = BH.oM.Geometry.Tolerance.Distance)
+        {
+            Point position = element0D.IGeometry().Project(Plane.XY);
+            ICurve gridCurve = grid.Curve.IProject(Plane.XY);
+
+            return position.IDistance(gridCurve) <= tolerance;
+        }
+
+        /******************************************/
+
+        [Description("Checks if the geometrical representation of an IElement1D, projected onto the XY Plane, is fully within a set tolerance from a grid line.")]
+        [Input("element1D", "The IElement1D that will be checked for proximity to the grid.")]
+        [Input("grid", "The Grid to use for evaulation.")]
+        [Input("tolerance", "The maximum distance allowed from the Grid for this method to return true.", typeof(Length))]
+        [Output("isFullyOnGrid", "A boolean which is true if the geometrical representation of an IElement1D is fully within a set tolerance from the grid line.")]
+        public static bool IsFullyOnGrid(this IElement1D element1D, Grid grid, double tolerance = BH.oM.Geometry.Tolerance.Distance)
+        {
+            ICurve curve = element1D.IGeometry().IProject(Plane.XY);
+            List <Point> ctrlPts = curve.ControlPoints();
+
+            ICurve gridCurve = grid.Curve.IProject(Plane.XY);
+
+            // All this method does is check if the control points are within tolerance from the grid, not if the actual curve is.
+            // This works well for Lines and Arcs, but to properly support NurbsCurves a more complex algorithm would be required,
+            // as control points of NurbsCurves are not usually on the curve itself. Simply using ClosestPoint to bring the control 
+            // points onto the curve would not be enough, as they may often end up in completely the wrong place and may not accurately
+            // reflect local maximi in distance from the curve. ClosestPoint does not yet support NurbsCurves anyway.
+            if (curve.IsNurbsCurve())
+            {
+                BH.Engine.Reflection.Compute.RecordWarning("IsFullyOnGrid does not fully support NurbsCurves. Results may be inaccurate.");
+            }
+
+            foreach (Point pt in ctrlPts)
+            {
+                if (pt.IDistance(gridCurve) >= tolerance)
+                    return false;
+            }
+
+            return ctrlPts.Count > 0;
```

Comment:
Won't this will always be true? And might thus be easier to just do `return true;`?

Reply:
You are probably right. It's just I've been working on null checks, and that made me wonder what would happen if you fed it an empty curve object with no curve segments. Since I didn't want it to return true in that case or some other when there are no control points returned, I added this check. But I don't think there's currently a way to trigger this that doesn't just crash in one of one of the earlier methods.